content/number-theory/Eratosthenes.h

```diff
@@ -0,0 +1,36 @@
+/**
+ * Author: Jakob Kogler, chilli, pajenegod
+ * Date: 2020-04-12
+ * License: CC0
+ * Description: Prime sieve for generating all primes up to LIM.
+ * Status: Stress-tested
+ * Time: LIM=1e9 $\approx$ 1.5s
+ * Details: Despite its n log log n complexity, segmented sieve is still faster
+ * than other options, including bitset sieves and linear sieves. This is
+ * primarily due to its low memory usage, which reduces cache misses. This
+ * implementation skips even numbers.
+ *
+ * Benchmark can be found here: https://ideone.com/e7TbX4
+ */
+#pragma once
+
+const int LIM = 1e6;
+bitset<LIM> isPrime;
+vi eratosthenes() {
+	const int S = round(sqrt(LIM)), R = LIM / 2;
+	vi pr({2}), sieve(S + 1); pr.reserve(LIM / (int)log(LIM));
```

Comment:
If you want to reserve the memory for `pr`, you probably should use a bigger value. The actual number of primes is a little bigger than `LIM / log(LIM)`, e.g.  ~6% for `LIM = 10^8`. So this still requires the vector to resize once, ending with almost twice the needed memory. For reference: https://en.wikipedia.org/wiki/Prime-counting_function#Table_of_%CF%80(x),_x_/_ln_x,_and_li(x)
Also, shouldn't the conversion to `int` happen after the division?

I suggest something like `(int)(LIM / log(LIM) * 1.1)` which fits for all `LIM >= 10^5`. Or `(int)(LIM / log(LIM) * 1.07` for all `LIM >= 10^7)`. (For smaller `LIM` the vector of primes is small anyhow, and it doesn't matter how much memory we use.

Reply:
Good point - I was just being lazy. Considering that this is needed to squeeze the memory to `200MB` though, I think this is a good idea.